src/uct/ib/rdmacm/rdmacm_ep.c

```diff
@@ -31,20 +32,50 @@ static UCS_CLASS_INIT_FUNC(uct_rdmacm_ep_t, uct_iface_t *tl_iface,
     self->conn_reply_arg = arg;
     self->priv_data      = ucs_malloc(sizeof(hdr) + length, "client private data");
     if (self->priv_data == NULL) {
-        return UCS_ERR_NO_MEMORY;
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
     }
 
     memcpy(self->priv_data, &hdr, sizeof(hdr));
     memcpy(self->priv_data + sizeof(hdr), priv_data, length);
 
+    /* TODO When UCT_CB_FLAG_SYNC is supported, return UCS_INPROGRESS since for
+     * librdmacm the reply_cb still needs to be invoked and it will be from the
+     * main thread after getting a reply from the server side */
+    ucs_assertv_always((cb_flags & UCT_CB_FLAG_ASYNC), "UCT_CB_FLAG_SYNC is not supported");
+
+    /* If the user's callbacks are called from the async thread, we cannot
+     * tell if by the end of this function they were already invoked and if this
+     * client's ep would already be connected to the server */
+    self->cb_flags = cb_flags;
+
     /* The interface can point at one endpoint at a time and therefore, the
      * connection establishment cannot be done in parallel for several endpoints */
     /* TODO support connection establishment on parallel endpoints on the same iface */
     if (iface->ep == NULL) {
         iface->ep = self;
+        self->is_on_pending = 0;
+
+        /* After rdma_resolve_addr(), the client will wait for an
+         * RDMA_CM_EVENT_ADDR_RESOLVED event on the event_channel
+         * to proceed with the connection establishment.
+         * This event will be retrieved from the event_channel by the async thread.
+         * All endpoints share the interface's event_channel but can use it serially. */
+         if (rdma_resolve_addr(iface->cm_id, NULL, (struct sockaddr *)sockaddr->addr,
```

Comment:
need to unite code with uct_rdmacm_iface_client_start_next_ep() and maintain fairness between eps:
always put new ep on the pending list at first, and then execute the pending queue in order, or schedule a slow-path progress to execute it

Reply:
@yosefe which part should be united? in this function the ep is only created so it isn't set to failed if it fails, we just exit the funcion.
an ep can be added to the list only from here, why isn't faireness maintained?

Reply:
1. rdma_resolve_addr should be called only from one place
2. re fairness - looks like you are right